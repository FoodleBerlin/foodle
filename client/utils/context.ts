import jwt from 'jsonwebtoken';

export const extractUserFromToken = (rawCookieToken: string | null, rawHeaderToken: string): any | null => {
  if (!rawHeaderToken && !rawCookieToken) return null;
  const decoded = jwt.verify(
    rawCookieToken ?? rawHeaderToken,
    process.env.SERVER_SECRET ?? '',
    (err: any, data: any) => {
      if (!err && data.user.id) {
        return data.user;
      } else {
        return null;
      }
    }
  );
  // JWT verify lib functionr returns void
  return decoded as unknown as any;
};
