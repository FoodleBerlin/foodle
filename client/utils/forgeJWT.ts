
export type Token = {
  user: {
    id: string;
    fullName: string;
    email: string;
    stripeId: string | null;
  };
};

