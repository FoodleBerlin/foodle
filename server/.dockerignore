*.png
*.md
*.jpg
#node_modules/ # Some issues appear if this one is not commented out, idk why exactly. TODO: Investigate at later on
tests/