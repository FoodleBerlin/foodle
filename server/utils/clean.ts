import prisma from '../singletons/prisma';

export async function clean() {
  try {
    await prisma.$executeRaw`DELETE FROM "User";`;
  } catch (e) {
    throw new Error(`Unable to seed ${e}`);
  }
}
