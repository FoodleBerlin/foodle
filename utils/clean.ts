import prisma from "../server/singletons/prisma";
export async function clean() {
  try {
    await prisma.$executeRaw`DELETE FROM "Booking";`;
    await prisma.$executeRaw`DELETE FROM "Property";`;
    await prisma.$executeRaw`DELETE FROM "User";`;
  } catch (e) {
    throw new Error(`Unable to seed ${e}`);
  }
}
