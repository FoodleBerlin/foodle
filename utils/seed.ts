import prisma from '../server/singletons/prisma';
export async function seed() {
  const [users] = await Promise.all(
    await prisma.$transaction([
      prisma.user.createMany({
        data: [
          {
            id: '1',
            email: 'user1@gmail.com',
            handle: 'user1',
            fullName: 'User 1',
            role: 'tenant',
            zip: 13000,
          },
          {
            id: '2',
            email: 'use2@gmail.com',
            handle: 'user2',
            fullName: 'User 2',
            role: 'tenant',
            zip: 14000,
          },
          {
            id: '3',
            email: 'user3@gmail.com',
            handle: 'user3',
            fullName: 'User 3',
            role: 'tenant',
            zip: 14000,
          },
          {
            id: '4',
            email: 'user4@gmail.com',
            handle: 'user4',
            fullName: 'User 4',
            role: 'landlord',
            zip: 15000,
          },
          {
            id: '5',
            email: 'user5@gmail.com',
            handle: 'user5',
            fullName: 'User 5',
            role: 'tenantLandlord',
            zip: 16000,
          },
        ],
      }),
    ])
  );
  await prisma.property.createMany({
    data: [
      {
        id: '1',
        size: 123,
        ownerId: '1',
        street: 'testhbdbjkfsdbjf',
        streetNumber: 1233,
        zip: 123445,
        city: 'berlin',
        description: 'testhbdbjkfsdbjf',
        rules: 'fffff',
        dailyPrice: 100,
        serviceFee: 0,
      },
    ],
  });
}
export default seed;
