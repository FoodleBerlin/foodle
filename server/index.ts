import { ApolloServer } from 'apollo-server-express';
import schema from './schema';
import { createContext } from './context';
import express from 'express';
import session from 'express-session';
import passport from './passport';
import forgeJWT from '../utils/forgeJWT';
import { DataSources } from 'apollo-server-core/dist/graphqlOptions';
import StripeWrapper from './singletons/stripe/endpoints';
import datasources from './singletons/datasources';

const port = process.env.PORT || 5000

export const app = express();
app.use(passport.initialize());

app.use(
  session({
    secret: process.env.SERVER_SECRET ?? '',
    resave: false,
    saveUninitialized: false,
  })
);
export const apollo: ApolloServer = new ApolloServer({
  // An executable GraphQL schema.
  schema,
  // An object (or a function that creates an object) that's passed to every resolver that executes for a particular operation.
  introspection: true,
  // This enables resolvers to share helpful context, such as a database connection.
  context: createContext,
  dataSources: () => datasources() as DataSources<Record<'stripeWrapper', StripeWrapper>>,
});

export const router = express.Router();

export async function main() {
  await apollo.start();
  app.use(router);
  apollo.applyMiddleware({ app });
  app.listen({
    port: port,
  });
}
if (!process.env.TEST) {
  main();
}

router.get('/api/auth', passport.authenticate('google', { scope: ['profile', 'email'] }));

router.get('/api/callback', (req: any, res: any, next) => {
  passport.authenticate('google', async (err: any, user: any) => {
    const token = await forgeJWT(user);
    res.cookie('jwt', token, {
      httpOnly: true,
      secure: false, // true in prod,
      sameSite: 'lax', // 'strict' in prod,
    });
    return res.redirect(process.env.CLIENT_URL);
  })(req, res, next);
});
