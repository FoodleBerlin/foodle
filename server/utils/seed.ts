import prisma from '../singletons/prisma';
import { clean } from './clean';
export async function seed() {
  await clean();
  const [users] = await Promise.all(
    await prisma.$transaction([
      prisma.user.createMany({
        data: [
          {
            id: '1',
            email: 'user1@gmail.com',
            handle: 'user1',
            fullName: 'User 1',
            role: 'tenant',
            zip: 13000,
            description: 'desc',
            passportS3Id: 'passS3',
            solvencyS3Id: 'solvS3',
            licenseS3Id: 'liceS3',
          },
          {
            id: '2',
            email: 'use2@gmail.com',
            handle: 'user2',
            fullName: 'User 2',
            role: 'tenant',
            zip: 14000,
            description: 'desc',
            passportS3Id: 'passS3',
            solvencyS3Id: 'solvS3',
            licenseS3Id: 'liceS3',
          },
          {
            id: '3',
            email: 'user3@gmail.com',
            handle: 'user3',
            fullName: 'User 3',
            role: 'tenant',
            zip: 14000,
            description: 'desc',
            passportS3Id: 'passS3',
            solvencyS3Id: 'solvS3',
            licenseS3Id: 'liceS3',
          },
          {
            id: '4',
            email: 'user4@gmail.com',
            handle: 'user4',
            fullName: 'User 4',
            role: 'landlord',
            zip: 15000,
            description: 'desc',
            passportS3Id: 'passS3',
            solvencyS3Id: 'solvS3',
            licenseS3Id: 'liceS3',
          },
          {
            id: '5',
            email: 'user5@gmail.com',
            handle: 'user5',
            fullName: 'User 5',
            role: 'tenantLandlord',
            zip: 16000,
            description: 'desc',
            passportS3Id: 'passS3',
            solvencyS3Id: 'solvS3',
            licenseS3Id: 'liceS3',
          },
        ],
      }),
    ])
  );

  await prisma.$transaction([
    prisma.property.createMany({
      data: [
        {
          zip: 1355,
          size: 123,
          street: 'sample-street',
          streetNumber: '23',
          city: 'Berlin',
          description: 'user1-listing',
          serviceFee: 5,
          rules: ['rule-1', 'rule-2'],
          title: 'prop title',
          hourlyPrice: 12,
          deposit: 2,
          images: ['test-image-url'],
          partialSpace: false,
          ownerId: '1',
          handle: 'prop1',
          frequency: 'weekly',
          facilities: ['Unfurnished'],
          startDate: '2022-06-27T08:00:00.003Z',
          endDate: '2022-06-30T08:00:00.003Z',
        },
        {
          zip: 1355,
          size: 123,
          street: 'sample-street',
          streetNumber: '23',
          city: 'Berlin',
          description: 'user1-listing',
          serviceFee: 5,
          rules: ['rule-1', 'rule-2'],
          title: 'prop title',
          hourlyPrice: 12,
          deposit: 2,
          images: ['test-image-url'],
          partialSpace: false,
          ownerId: '2',
          handle: 'prop2',
          frequency: 'weekly',
          facilities: ['Elevator', 'Parking', 'Dishwasher'],
          startDate: '2022-08-27T08:00:00.003Z',
          endDate: '2022-09-30T08:00:00.003Z',
        },
        {
          zip: 1355,
          size: 123,
          street: 'sample-street',
          streetNumber: '23',
          city: 'Berlin',
          description: 'user1-listing',
          serviceFee: 5,
          rules: ['rule-1', 'rule-2'],
          title: 'prop title',
          hourlyPrice: 12,
          deposit: 2,
          images: ['test-image-url'],
          partialSpace: false,
          ownerId: '2',
          handle: 'prop3',
          frequency: 'weekly',
          facilities: ['Elevator', 'Parking'],
          startDate: '2022-10-27T08:00:00.003Z',
          endDate: '2022-11-30T08:00:00.003Z',
        },
      ],
    }),
  ]);

  const prop1 = await prisma.property.findUnique({
    where: {
      handle: 'prop1',
    },
  });
  if (prop1 == null) {
    throw Error('Error while seeding db, property with handle prop1 does not exist.');
  }
  const prop2 = await prisma.property.findUnique({
    where: {
      handle: 'prop2',
    },
  });
  if (prop2 == null) {
    throw Error('Error while seeding db, property with handle prop2 does not exist.');
  }
  // seeding for booking request
  /*  
      test weekly, one weekday
      {
        daySlots: {
          startTime: '2022-06-27T08:00:00.003Z',
          endTime: '2022-06-27T16:00:00.003Z',
        },
        frequency: 'NONE',
        startDate: '2022-06-27T08:00:00.003Z',
        endDate: '2022-07-25T08:00:00.003Z',
        propertyHandle: 'prop1',
      } */
  await prisma.$transaction([
    prisma.daySlot.createMany({
      data: [
        {
          startTime: '2022-06-27T08:00:00.003Z',
          endTime: '2022-06-27T16:00:00.003Z',
          propertyId: prop1.id,
        },
        {
          startTime: '2022-07-04T08:00:00.003Z',
          endTime: '2022-07-04T16:00:00.003Z',
          propertyId: prop1.id,
        },
        {
          startTime: '2022-07-11T08:00:00.003Z',
          endTime: '2022-07-11T16:00:00.003Z',
          propertyId: prop1.id,
        },
        {
          startTime: '2022-07-18T08:00:00.003Z',
          endTime: '2022-07-18T16:00:00.003Z',
          propertyId: prop1.id,
        },
        {
          startTime: '2022-07-25T08:00:00.003Z',
          endTime: '2022-07-25T16:00:00.003Z',
          propertyId: prop1.id,
        },

        /* 
          test monthly + multiple weekdays + varying day time
          
          const bookingVars = {
            daySlots: [
              {
                startTime: '2022-05-23T10:00:00.003Z',
                endTime: '2022-05-23T16:00:00.003Z',
              },
              {
                startTime: '2022-05-25T08:00:00.003Z',
                endTime: '2022-05-25T16:00:00.003Z',
              },
            ],
            frequency: 'MONTHLY',
            startDate: '2022-05-25T08:00:00.003Z',
            endDate: '2022-07-22T08:00:00.003Z',
            propertyHandle: 'prop2',
          };
        */

        {
          startTime: '2022-05-25T08:00:00.003Z',
          endTime: '2022-05-25T16:00:00.003Z',
          propertyId: prop2.id,
        },
        {
          startTime: '2022-05-23T08:00:00.003Z',
          endTime: '2022-05-23T16:00:00.003Z',
          propertyId: prop2.id,
        },
        {
          startTime: '2022-06-20T08:30:00.003Z',
          endTime: '2022-06-20T16:00:00.003Z',
          propertyId: prop2.id,
        },
        {
          startTime: '2022-06-22T08:00:00.003Z',
          endTime: '2022-06-22T16:00:00.003Z',
          propertyId: prop2.id,
        },
        {
          startTime: '2022-07-18T09:00:00.003Z',
          endTime: '2022-07-18T16:00:00.003Z',
          propertyId: prop2.id,
        },
        {
          startTime: '2022-07-20T08:00:00.003Z',
          endTime: '2022-07-20T16:00:00.003Z',
          propertyId: prop2.id,
        },
        /* 
          test frequency none
          {
          "daySlots": {
            "startTime": "2022-06-27T09:00:00.003Z",
            "endTime": "2022-06-27T13:00:00.003Z",
            "weekday": "FRI"
          },
           {
            "startTime": "2022-06-27T08:00:00.003Z",
            "endTime": "2022-06-27T16:00:00.003Z",
            "weekday": "WED"
          },
          "frequency": "NONE",
          "startDate": "2022-05-25T08:00:00.003Z",
          "endDate": "2022-07-22T08:00:00.003Z",
          "propertyHandle": "prop2"
        } */

        {
          startTime: '2022-08-19T08:00:00.003Z',
          endTime: '2022-08-19T16:00:00.003Z',
          propertyId: prop2.id,
        },
      ],
    }),
  ]);
}
