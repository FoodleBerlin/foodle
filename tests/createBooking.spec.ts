import { clean } from '../utils/clean';
import { seed } from '../utils/seed';
import { apollo as server } from './server/index';

beforeAll(async () => {
  await clean();
  await seed();
});

const query = `mutation CreateBooking($daySlots: [AvailableDay!]!, $frequency: FrequencyEnum!, $endDate: DateTime!, $startDate: DateTime!, $propertyHandle: String!) {
    createBooking(daySlots: $daySlots, frequency: $frequency, endDate: $endDate, startDate: $startDate, propertyHandle: $propertyHandle) {
      Booking {
        tenant {
          fullName
          email
          handle
          zip
        }
        property {
          handle
          title
          size
        }
        bookingStatus
        frequency
        totalPrice
        startDate
        endDate
      }
      ClientErrorUserNotExists {
        message
      }
      ClientErrorPropertyNotExists {
        message
      }
      ClientErrorInvalidInput {
        message
      }
      NoAvailableSlots {
        message
      }
      ClientErrorInvalidPropertyInput {
        message
      }
      UnknownError {
        message
      }
    }
  }`;

describe('Create Booking on Listing success with ', () => {
  it('frequency weekly, one weekday, on property proeprty2', async () => {
    const bookingVars = {
      daySlots: {
        startTime: '2022-06-27T08:00:00.003Z',
        endTime: '2022-06-27T16:00:00.003Z',
      },
      frequency: 'WEEKLY',
      startDate: '2022-06-27T08:00:00.003Z',
      endDate: '2022-07-25T08:00:00.003Z',
      propertyHandle: 'prop1',
    };
    const res = await server.executeOperation({
      query,
      variables: {
        ...bookingVars,
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('frequency monthly, multiple weekdays, varying day time, property 2', async () => {
    const bookingVars = {
      daySlots: [
        {
          startTime: '2022-05-23T10:00:00.003Z',
          endTime: '2022-05-23T16:00:00.003Z',
        },
        {
          startTime: '2022-05-25T08:00:00.003Z',
          endTime: '2022-05-25T16:00:00.003Z',
        },
      ],
      frequency: 'MONTHLY',
      startDate: '2022-05-25T08:00:00.003Z',
      endDate: '2022-07-22T08:00:00.003Z',
      propertyHandle: 'prop2',
    };
    const res = await server.executeOperation({
      query,
      variables: {
        ...bookingVars,
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('frequency none, one weekday, second booking on one listing, property 2', async () => {
    const bookingVars = {
      daySlots: [
        {
          startTime: '2022-06-27T09:00:00.003Z',
          endTime: '2022-06-27T13:00:00.003Z',
        },
        {
          startTime: '2022-06-27T08:00:00.003Z',
          endTime: '2022-06-27T16:00:00.003Z',
        },
      ],
      frequency: 'NONE',
      startDate: '2022-05-25T08:00:00.003Z',
      endDate: '2022-07-22T08:00:00.003Z',
      propertyHandle: 'prop2',
    };
    const res = await server.executeOperation({
      query,
      variables: {
        ...bookingVars,
      },
    });
    expect(res).toMatchSnapshot();
  });
});
