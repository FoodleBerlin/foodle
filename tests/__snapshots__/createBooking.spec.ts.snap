// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Booking on Listing success with  frequency weekly, one weekday, on property proeprty2 1`] = `
Object {
  "data": Object {
    "createBooking": Object {
      "Booking": Object {
        "bookingStatus": "PENDING",
        "endDate": "1658736000003",
        "frequency": "WEEKLY",
        "property": Object {
          "handle": "prop1",
          "size": 123,
          "title": "prop title",
        },
        "startDate": "1656316800003",
        "tenant": Object {
          "email": "user1@gmail.com",
          "fullName": "User 1",
          "handle": "user1",
          "zip": 13000,
        },
        "totalPrice": 384,
      },
      "ClientErrorInvalidInput": null,
      "ClientErrorInvalidPropertyInput": null,
      "ClientErrorPropertyNotExists": null,
      "ClientErrorUserNotExists": null,
      "NoAvailableSlots": null,
      "UnknownError": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
